flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java

```diff
@@ -181,86 +168,22 @@ private static ConstantObjectInspector getPrimitiveJavaConstantObjectInspector(P
-				if (((JavaLongObjectInspector) inspector).preferWritable()) {
-					return o -> new LongWritable((Long) o);
-				} else {
-					return IdentityConversion.INSTANCE;
-				}
-			} else if (inspector instanceof JavaFloatObjectInspector) {
-				if (((JavaFloatObjectInspector) inspector).preferWritable()) {
-					return o -> new FloatWritable((Float) o);
-				} else {
-					return IdentityConversion.INSTANCE;
-				}
-			} else if (inspector instanceof JavaDoubleObjectInspector) {
-				if (((JavaDoubleObjectInspector) inspector).preferWritable()) {
-					return o -> new DoubleWritable((Double) o);
-				} else {
-					return IdentityConversion.INSTANCE;
-				}
-			} else if (inspector instanceof JavaDateObjectInspector) {
-				if (((JavaDateObjectInspector) inspector).preferWritable()) {
-					return o -> new DateWritable((Date) o);
-				} else {
-					return IdentityConversion.INSTANCE;
-				}
-			} else if (inspector instanceof JavaTimestampObjectInspector) {
-				if (((JavaTimestampObjectInspector) inspector).preferWritable()) {
-					return o -> new TimestampWritable((Timestamp) o);
-				} else {
-					return IdentityConversion.INSTANCE;
-				}
-			} else if (inspector instanceof JavaBinaryObjectInspector) {
-				if (((JavaBinaryObjectInspector) inspector).preferWritable()) {
-					return o -> new BytesWritable((byte[]) o);
-				} else {
-					return IdentityConversion.INSTANCE;
-				}
-			} else if (inspector instanceof JavaHiveCharObjectInspector) {
-				if (((JavaHiveCharObjectInspector) inspector).preferWritable()) {
-					return o -> new HiveCharWritable(
-						new HiveChar((String) o, ((CharType) dataType.getLogicalType()).getLength()));
-				} else {
-					return o -> new HiveChar((String) o, ((CharType) dataType.getLogicalType()).getLength());
-				}
-			} else if (inspector instanceof JavaHiveVarcharObjectInspector) {
-				if (((JavaHiveVarcharObjectInspector) inspector).preferWritable()) {
-					return o -> new HiveVarcharWritable(
-						new HiveVarchar((String) o, ((VarCharType) dataType.getLogicalType()).getLength()));
-				} else {
-					return o -> new HiveVarchar((String) o, ((VarCharType) dataType.getLogicalType()).getLength());
-				}
+			if (inspector instanceof BooleanObjectInspector ||
+					inspector instanceof StringObjectInspector ||
+					inspector instanceof ByteObjectInspector ||
+					inspector instanceof ShortObjectInspector ||
+					inspector instanceof IntObjectInspector ||
+					inspector instanceof LongObjectInspector ||
+					inspector instanceof FloatObjectInspector ||
+					inspector instanceof DoubleObjectInspector ||
+					inspector instanceof DateObjectInspector ||
+					inspector instanceof TimestampObjectInspector ||
+					inspector instanceof BinaryObjectInspector) {
+				return IdentityConversion.INSTANCE;
+			} else if (inspector instanceof HiveCharObjectInspector) {
+				return o -> new HiveChar((String) o, ((CharType) dataType.getLogicalType()).getLength());
```

Comment:
Here we are converting Flink data to Hive data, and it seems that we probably should use the precision specified by Hive (in HiveCharObjectInspector) instead. The same for HiveVarchar. This might make a difference in certain cases.

Reply:
the precision in HiveChar/VarcharObjectInspector here is actually set from the LogicalType, thus Flink's LogicalType is the single source of truth

Reply:
I don't think this is about a single source of truth. It's a data conversion. We are converting data from Flink to Hive. While data type is char on both end, they may come with different precision. For instance, the data could be char(10), while hive's data type is char(5). During the conversion, the data should be enforced with precision 5, not 10.

Reply:
discussed with Xuefu offline, we will leave this as-is